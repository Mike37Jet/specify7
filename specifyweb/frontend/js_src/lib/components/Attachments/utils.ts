import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import { getModel, schema } from '../DataModel/schema';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyModel } from '../DataModel/specifyModel';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(schema.models).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getModel(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);
/** Exclude tables without read access*/
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyModel
): Relationship | undefined =>
  table.name === 'Attachment'
    ? undefined
    : table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedModel.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      });
