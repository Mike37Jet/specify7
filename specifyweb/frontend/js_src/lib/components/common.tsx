/**
 * Generic React Components
 *
 * @module
 *
 */

import React from 'react';
import ReactDOM from 'react-dom';

import { f } from '../functools';
import { spanNumber } from '../helpers';
import { getIcon } from '../icons';
import { commonText } from '../localization/common';
import { getModel } from '../schema';
import { Button, className, Textarea } from './basic';
import { copyTextToClipboard } from './filepicker';
import { useBooleanState, useTitle } from './hooks';
import { icons } from './icons';
import { compareStrings } from './internationalization';
import { usePref } from './preferenceshooks';

const MAX_HUE = 360;

/**
 * Convert first 2 characters of a table name to a number [0,255] corresponding
 * to color hue.
 *
 * Used for autogenerated table icons if table icon image is missing.
 */
const getHue = spanNumber(
  // eslint-disable-next-line unicorn/prefer-code-point
  'a'.charCodeAt(0) * 2,
  // eslint-disable-next-line unicorn/prefer-code-point
  'z'.charCodeAt(0) * 2,
  0,
  MAX_HUE
);

/** Generate an HSL color based on the first 2 characters of a string */
export const stringToColor = (name: string) =>
  f.var(
    name.toLowerCase(),
    (name) =>
      `hsl(${getHue(
        // eslint-disable-next-line unicorn/prefer-code-point
        (name[0] ?? 'a').charCodeAt(0) + (name[1] ?? 'a').charCodeAt(0)
      )}, 70%, 50%)`
  );

/**
 * Renders a table icon or autogenerates a new one
 */
export function TableIcon({
  name,
  tableLabel,
  /*
   * It is highly recommended to use the same icon size everywhere, as that
   * improves consistency, thus, this should be overwritten only if it is
   * strictly necessary.
   */
  className = 'w-table-icon h-table-icon',
}: {
  readonly name: string;
  readonly tableLabel?: string | false;
  readonly className?: string;
}): JSX.Element {
  const tableIconSource = getIcon(name);
  const resolvedTableLabel =
    tableLabel === false
      ? undefined
      : tableLabel ?? getModel(name)?.label ?? '';
  const role = typeof resolvedTableLabel === 'string' ? 'img' : undefined;
  const ariaHidden = resolvedTableLabel === undefined;
  if (typeof tableIconSource === 'string')
    return (
      <span
        className={`${className} bg-center bg-no-repeat bg-contain`}
        role={role}
        style={{ backgroundImage: `url('${tableIconSource}')` }}
        title={resolvedTableLabel}
        aria-label={typeof role === 'string' ? resolvedTableLabel : undefined}
        aria-hidden={ariaHidden}
      />
    );

  // If icon is missing, show an autogenerated one:
  return (
    <span
      style={{ backgroundColor: stringToColor(name) }}
      role={role}
      className={`w-table-icon h-table-icon flex items-center justify-center
        text-white rounded-sm text-sm`}
      title={resolvedTableLabel}
      aria-label={resolvedTableLabel}
      aria-hidden={ariaHidden}
    >
      {name.slice(0, 2).toUpperCase()}
    </span>
  );
}

export const tableIconUndefined = (
  <span
    className="w-table-icon h-table-icon flex items-center justify-center font-bold text-red-600"
    aria-label={commonText('unmapped')}
    role="img"
  >
    {icons.ban}
  </span>
);

export const tableIconSelected = (
  <span
    className="w-table-icon h-table-icon flex items-center justify-center font-bold text-green-500"
    aria-label={commonText('mapped')}
    role="img"
  >
    {icons.check}
  </span>
);

export const tableIconEmpty = (
  <span className="w-table-icon h-table-icon" aria-hidden={true} />
);

/** Internationalized bi-directional string comparison function */
export const compareValues = (
  ascending: boolean,
  valueLeft: string | undefined,
  valueRight: string | undefined
): number =>
  compareStrings(valueLeft ?? '', valueRight ?? '') * (ascending ? -1 : 1);

export type SortConfig<FIELD_NAMES extends string> = {
  readonly sortField: FIELD_NAMES;
  readonly ascending: boolean;
};

export function SortIndicator<FIELD_NAMES extends string>({
  fieldName,
  sortConfig,
}: {
  readonly fieldName: string;
  readonly sortConfig: SortConfig<FIELD_NAMES>;
}): JSX.Element {
  const isSorted = sortConfig.sortField === fieldName;
  return (
    <span className="text-brand-300">
      {isSorted && (
        <span className="sr-only">
          {sortConfig.ascending
            ? commonText('ascending')
            : commonText('descending')}
        </span>
      )}
      {isSorted
        ? sortConfig.ascending
          ? icons.chevronUp
          : icons.chevronDown
        : undefined}
    </span>
  );
}

/**
 * A React Portal wrapper
 *
 * @remarks
 * Based on https://blog.logrocket.com/learn-react-portals-by-example/
 *
 * Used when an elements needs to be renreded outside of the bounds of
 * the container that has overflow:hidden
 */
export function Portal({
  children,
}: {
  readonly children: JSX.Element;
}): JSX.Element {
  const element = React.useMemo(() => document.createElement('div'), []);

  React.useEffect(() => {
    const portalRoot = document.getElementById('portal-root');
    if (portalRoot === null) throw new Error('Portal root was not found');
    portalRoot.append(element);
    return (): void => element.remove();
  }, [element]);

  return ReactDOM.createPortal(children, element);
}

export function AppTitle({
  title,
  type,
}: {
  readonly title: string;
  readonly type?: 'form';
}): null {
  const [updateTitle] = usePref('form', 'behavior', 'updatePageTitle');
  useTitle(type !== 'form' || updateTitle ? title : undefined);
  return null;
}

export function AutoGrowTextArea({
  containerClassName,
  ...props
}: Parameters<typeof Textarea>[0] & {
  readonly containerClassName?: string;
}): JSX.Element {
  const [textArea, setTextArea] = React.useState<HTMLTextAreaElement | null>(
    null
  );
  const [shadow, setShadow] = React.useState<HTMLDivElement | null>(null);
  /*
   * If user manually resized the textarea, need to keep the shadow in sync
   * Fixes https://github.com/specify/specify7/issues/1783
   * Can't simply convert auto growing textarea into a regular one on the fly
   * because that interrupts the resize operation
   */
  React.useEffect(() => {
    if (
      textArea === null ||
      shadow === null ||
      globalThis.ResizeObserver === undefined
    )
      return undefined;
    const observer = new globalThis.ResizeObserver(() => {
      shadow.style.height = textArea.style.height;
      shadow.style.width = textArea.style.width;
    });
    observer.observe(textArea);
    return (): void => observer.disconnect();
  }, [textArea, shadow]);
  return (
    <div
      className={`
        relative min-h-[calc(theme(spacing.7)*var(--rows))] overflow-hidden
        ${containerClassName ?? ''}
      `}
      style={{ '--rows': props.rows ?? 3 } as React.CSSProperties}
    >
      {/*
       * Shadow a textarea with a div, allowing it to autoGrow. Source:
       * https://css-tricks.com/the-cleanest-trick-for-autogrowing-textareas/
       */}
      <div
        className={`
          textarea-shadow print:hidden invisible whitespace-pre-wrap
          [grid-area:1/1/2/2] ${className.textArea}
        `}
        ref={setShadow}
      >
        {`${props.value?.toString() ?? ''} `}
      </div>
      <Textarea
        {...props}
        className={`
          h-full top-0 [grid-area:1/1/2/2] absolute
          ${props.className ?? ''}
        `}
        forwardRef={(textArea): void => {
          setTextArea(textArea);
          if (typeof props.forwardRef === 'function')
            props.forwardRef(textArea);
          else if (
            typeof props.forwardRef === 'object' &&
            props.forwardRef !== null &&
            'current' in props.forwardRef
          )
            /* REFACTOR: improve typing to make this editable */
            // @ts-expect-error Modifying a read-only property
            props.forwardRef.current = textArea;
        }}
      />
    </div>
  );
}

const copyMessageTimeout = 3000;

export function CopyButton({
  text,
  label = commonText('copyToClipboard'),
}: {
  readonly text: string;
  readonly label?: string;
}): JSX.Element {
  const [wasCopied, handleCopied, handleNotCopied] = useBooleanState();
  return (
    <Button.Green
      className="whitespace-nowrap"
      onClick={(): void =>
        void copyTextToClipboard(text).then((): void => {
          handleCopied();
          globalThis.setTimeout(handleNotCopied, copyMessageTimeout);
        })
      }
    >
      {wasCopied ? commonText('copied') : label}
    </Button.Green>
  );
}
