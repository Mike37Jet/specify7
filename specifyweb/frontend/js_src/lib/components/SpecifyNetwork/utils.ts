import type { IR } from '../../utils/types';
import { R, RA } from '../../utils/types';
import { Field, LocalityData } from '../Leaflet/helpers';
import {
  deflateMappingPaths,
  getCanonicalMappingPath,
  mappingPathToString,
  splitJoinedMappingPath,
} from '../WbPlanView/mappingHelpers';
import { MappingPath } from '../WbPlanView/Mapper';
import { findArrayDivergencePoint } from '../../utils/utils';

const fieldOrder: IR<{ readonly after: string }> = {
  'gbif:acceptedScientificName': { after: 'dwc:scientificName' },
  'dwc:datasetName': { after: 'dwc:specificEpithet' },
};

export function reorderBrokerFields<T>(mappedTable: IR<T>): IR<T> {
  let reorderedTable = mappedTable;
  Object.entries(fieldOrder)
    .filter(
      ([fieldName, position]) =>
        !(fieldName in reorderedTable) ||
        !('after' in position) ||
        !(position.after in reorderedTable)
    )
    .forEach(([fieldName, { after }]) => {
      const table = Object.entries(reorderedTable);
      const keys = Object.keys(reorderedTable);
      const currentPosition = keys.indexOf(fieldName);
      let targetPosition = keys.indexOf(after);
      const currentEntry = table.splice(currentPosition, 1)[0];
      if (currentPosition > targetPosition) targetPosition += 1;

      table.splice(targetPosition, 0, currentEntry);
      reorderedTable = Object.fromEntries(table);
    });

  return reorderedTable;
}

export function deflateLocalityData(localityData: LocalityData): LocalityData {
  const deflatedMappingPaths = deflateMappingPaths(
    Object.keys(localityData).map(splitJoinedMappingPath)
  );
  return Object.fromEntries(
    Object.values(localityData).map((value, index) => [
      mappingPathToString(deflatedMappingPaths[index]),
      value,
    ])
  );
}

/*
 * Separates locality data for given -to-many mapping path into separate
 * locality data objects.
 *
 * For example, given this input:
 * localityData = {
 *   'locality.collectingevents.#2.collectionobjects.#1.catalognumber': 123,
 *   'locality.collectingevents.#3.collectionobjects.#4.catalognumber': 231,
 *   'locality.localityname': 'Lawrence',
 * }
 * filterMappingPath = ['locality','collectingevents','#1','collectionobjects']
 *
 * The result is:
 * [
 *   {
 *     'locality.collectingevents.#2.collectionobjects.#1.catalognumber': 123,
 *     'locality.localityname': 'Lawrence',
 *   }
 *   {
 *     'locality.collectingevents.#3.collectionobjects.#4.catalognumber': 231,
 *     'locality.localityname': 'Lawrence',
 *   }
 * ]
 */
export function splitLocalityData(
  localityData: LocalityData,
  filterMappingPath: MappingPath
): RA<LocalityData> {
  const filter = getCanonicalMappingPath(filterMappingPath);
  const groups = Object.entries(localityData).reduce<
    R<R<Field<number | string>>>
  >(
    (groups, [mappingPathString, field]) => {
      const mappingPath = splitJoinedMappingPath(mappingPathString);
      const canonicalMappingPath = getCanonicalMappingPath(mappingPath);
      const divergence = findArrayDivergencePoint(canonicalMappingPath, filter);
      const key =
        divergence === -1
          ? ''
          : mappingPathToString(mappingPath.slice(0, divergence));
      groups[key] ??= {};
      groups[key][mappingPathToString(mappingPath.slice(divergence))] = field;
      return groups;
    },
    { '': {} }
  );

  return Object.entries(groups)
    .filter(([key]) => key !== '')
    .map(([groupKey, groupData]) =>
      deflateLocalityData({
        ...groups[''],
        ...Object.fromEntries(
          Object.entries(groupData).map(([mappingPathString, field]) => [
            mappingPathToString([
              ...splitJoinedMappingPath(groupKey),
              ...splitJoinedMappingPath(mappingPathString),
            ]),
            field,
          ])
        ),
      })
    );
}
