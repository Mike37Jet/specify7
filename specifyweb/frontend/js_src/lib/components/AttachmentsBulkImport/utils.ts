import { attachmentsText } from '../../localization/attachments';
import { commonText } from '../../localization/common';
import { wbText } from '../../localization/workbench';
import { ajax } from '../../utils/ajax';
import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import { filterArray } from '../../utils/types';
import {
  insertItem,
  keysToLowerCase,
  mappedFind,
  replaceItem,
} from '../../utils/utils';
import { addMissingFields } from '../DataModel/addMissingFields';
import { deserializeResource, serializeResource } from '../DataModel/helpers';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { schema } from '../DataModel/schema';
import type { SpQuery, SpQueryField, Tables } from '../DataModel/types';
import type { UiFormatter } from '../Forms/uiFormatters';
import { queryFieldFilters } from '../QueryBuilder/FieldFilter';
import { makeQueryField } from '../QueryBuilder/fromTree';
import type { AttachmentUploadSpec } from './Import';
import { AttachmentMapping, staticAttachmentImportPaths } from './importPaths';
import type {
  AttachmentStatus,
  CanDelete,
  CanUpload,
  PartialUploadableFileSpec,
  UnBoundFile,
} from './types';
import { formatterTypeMapper } from '../Forms/uiFormatters';
import { State } from 'typesafe-reducer';
import { formsText } from '../../localization/forms';

const isAttachmentMatchValid = (uploadSpec: PartialUploadableFileSpec) =>
  uploadSpec.matchedId !== undefined &&
  (uploadSpec.matchedId.length === 1 || uploadSpec.disambiguated !== undefined);

export type ResolvedAttachmentRecord =
  | State<
      'invalid' | 'valid',
      { readonly reason: keyof typeof keyLocalizationMapAttachment }
    >
  | State<'matched', { readonly id: number }>;

const resolveAttachmentMatch = (
  matchedId: RA<number>,
  disambiguated: number | undefined
): ResolvedAttachmentRecord =>
  matchedId.length === 0
    ? { type: 'invalid', reason: 'noMatch' }
    : matchedId.length > 1 && disambiguated === undefined
    ? {
        type: 'invalid',
        reason: 'multipleMatches',
      }
    : {
        type: 'matched',
        id: disambiguated ?? matchedId[0],
      };
export function resolveAttachmentRecord(
  matchedId: RA<number> | undefined,
  disambiguated: number | undefined,
  parsedName: string | undefined
): ResolvedAttachmentRecord {
  if (parsedName === undefined)
    return { type: 'invalid', reason: 'incorrectFormatter' };
  if (matchedId === undefined)
    return { type: 'valid', reason: 'correctlyFormatted' };
  return resolveAttachmentMatch(matchedId, disambiguated);
}

const findFirstReason =
  (
    conditions: (uploadSpec: PartialUploadableFileSpec) => RA<boolean>,
    humanReasons: RA<keyof typeof keyLocalizationMapAttachment>
  ) =>
  (uploadSpec: PartialUploadableFileSpec) => {
    const resolvedConditions = conditions(uploadSpec);
    return mappedFind(resolvedConditions, (condition, index) =>
      condition ? undefined : humanReasons[index]
    );
  };

const uploadHumanReasons: RA<keyof typeof keyLocalizationMapAttachment> = [
  'alreadyUploaded',
  'noFile',
  'incorrectFormatter',
  'matchError',
];

const deleteHumanReasons: RA<keyof typeof keyLocalizationMapAttachment> = [
  'noAttachments',
  'matchError',
];
export const reasonToSkipUpload = findFirstReason(
  (uploadSpec) => [
    uploadSpec.attachmentId === undefined,
    uploadSpec.file instanceof File,
    uploadSpec.file.parsedName !== undefined,
    isAttachmentMatchValid(uploadSpec),
  ],
  uploadHumanReasons
);

export const reasonToSkipDelete = findFirstReason(
  (uploadSpec) => [
    uploadSpec.attachmentId !== undefined,
    isAttachmentMatchValid(uploadSpec),
  ],
  deleteHumanReasons
);

export const canUploadAttachment = (
  uploadSpec: PartialUploadableFileSpec
): uploadSpec is CanUpload => reasonToSkipUpload(uploadSpec) === undefined;

export const canDeleteAttachment = (
  uploadSpec: PartialUploadableFileSpec
): uploadSpec is CanDelete =>
  uploadSpec.attachmentId !== undefined && isAttachmentMatchValid(uploadSpec);

function generateInQueryResource(
  baseTable: keyof Tables,
  path: string,
  searchableList: RA<number | string | undefined>,
  queryName: string,
  additionalPaths: RA<
    Partial<SerializedResource<SpQueryField>> & { readonly path: string }
  > = []
): SpecifyResource<SpQuery> {
  const rawFields = [
    ...additionalPaths,
    {
      path,
      isDisplay: true,
      operStart: queryFieldFilters.in.id,
      startValue: filterArray(searchableList).join(','),
    },
  ];
  const queryField = rawFields.map(({ path, ...field }, index) =>
    serializeResource(
      makeQueryField(baseTable, path, { ...field, position: index })
    )
  );
  return deserializeResource(
    addMissingFields('SpQuery', {
      name: queryName,
      contextName: baseTable,
      contextTableId: schema.models[baseTable].tableId,
      countOnly: false,
      fields: queryField,
    })
  );
}

export async function validateAttachmentFiles(
  uploadableFiles: RA<PartialUploadableFileSpec>,
  uploadSpec: AttachmentUploadSpec,
  keepDisambiguation: boolean = false
): Promise<RA<PartialUploadableFileSpec>> {
  const { baseTable, path } =
    staticAttachmentImportPaths[uploadSpec.staticPathKey];

  const validationQueryResource = generateInQueryResource(
    baseTable,
    path,
    uploadableFiles.map(({ file }) => file?.parsedName),
    'Batch Attachment Upload'
  );
  const validationResponse = (
    await validationPromiseGenerator(validationQueryResource)
  )
    .map(({ targetId, restResult }) => ({
      targetId,
      rawResult: uploadSpec.formatQueryResults(restResult[0]),
    }))
    .filter((result) => result.rawResult !== undefined);

  return matchFileSpec(
    uploadableFiles,
    validationResponse as RA<{
      readonly targetId: number;
      readonly rawResult: string;
    }>,
    keepDisambiguation
  );
}

export function matchSelectedFiles(
  previousUploadables: RA<PartialUploadableFileSpec>,
  filesToResolve: RA<PartialUploadableFileSpec>
): RA<PartialUploadableFileSpec> {
  let resolvedFiles = previousUploadables;
  filesToResolve.forEach((uploadable) => {
    const matchedIndex = resolvedFiles.findIndex((previousUploadable) => {
      if (
        previousUploadable.attachmentId !== undefined ||
        previousUploadable.file instanceof File
      )
        return false;

      return (
        previousUploadable.file !== undefined &&
        previousUploadable.file.name === uploadable.file.name &&
        previousUploadable.file.size === uploadable.file.size &&
        previousUploadable.file.type === uploadable.file.type
      );
    });

    resolvedFiles =
      matchedIndex === -1
        ? insertItem(resolvedFiles, resolvedFiles.length, uploadable)
        : replaceItem(resolvedFiles, matchedIndex, {
            ...resolvedFiles[matchedIndex],
            file: uploadable.file,
            /*
             * Generating tokens again because the file could have been
             * uploaded to the asset server but not yet recorded in Specify DB.
             */
            uploadTokenSpec: undefined,
            // Take the new status in case of parse failure was reported.
            status: uploadable.status,
          });
  });
  return resolvedFiles;
}

export function resolveFileNames(
  previousFile: UnBoundFile,
  getFormatted: (
    rawName: number | string | null | undefined
  ) => string | undefined,
  formatter?: UiFormatter
): PartialUploadableFileSpec {
  let nameToParse: string | undefined;
  if (
    formatter === undefined ||
    formatter.fields.some((field) => field instanceof formatterTypeMapper.regex)
  )
    nameToParse = stripLastOccurrence(previousFile.name, '.');
  else {
    const formattedLength = formatter.fields.reduce(
      (length, field) => length + field.size,
      0
    );
    nameToParse = previousFile.name.slice(0, formattedLength);
  }
  previousFile.parsedName = f.maybe(nameToParse, getFormatted);

  return {
    file: previousFile,
  };
}

export function stripLastOccurrence(target: string, delimiter: string) {
  const splittedString = target.split(delimiter);
  return splittedString
    .slice(0, splittedString.length === 1 ? 1 : -1)
    .join(delimiter);
}

const validationPromiseGenerator = async (
  queryResource: SpecifyResource<SpQuery>
): Promise<
  RA<{
    readonly targetId: number;
    readonly restResult: RA<number | string>;
  }>
> =>
  ajax<{
    // First value is the primary key
    readonly results: RA<readonly [number, ...RA<number | string>]>;
  }>('/stored_query/ephemeral/', {
    method: 'POST',
    headers: {
      // eslint-disable-next-line @typescript-eslint/naming-convention
      Accept: 'application/json',
    },
    body: keysToLowerCase({
      ...serializeResource(queryResource),
      countOnly: false,
      limit: 0,
    }),
  }).then(({ data }) =>
    data.results.map(([target, ...restResult]) => ({
      targetId: target,
      restResult,
    }))
  );

const matchFileSpec = (
  uploadFileSpec: RA<PartialUploadableFileSpec>,
  queryResults: RA<{
    readonly targetId: number;
    readonly rawResult: string;
  }>,
  keepDisambiguation: boolean = false
): RA<PartialUploadableFileSpec> =>
  uploadFileSpec.map((spec) => {
    const specParsedName = spec.file?.parsedName;
    // Don't match files already uploaded.
    if (specParsedName === undefined || typeof spec.attachmentId === 'number')
      return spec;
    const matchingResults = queryResults.filter(
      (result) => result.rawResult === specParsedName
    );
    const newSpec: PartialUploadableFileSpec = {
      ...spec,
      matchedId: matchingResults.map((result) => result.targetId),
    };

    if (
      keepDisambiguation &&
      spec.disambiguated !== undefined &&
      // If disambiguation was chosen, but it became invalid, reset disambiguation
      newSpec.matchedId?.includes(spec.disambiguated)
    ) {
      return newSpec;
    }
    return { ...newSpec, disambiguated: undefined };
  });

export async function reconstructDeletingAttachment(
  staticKey: keyof typeof staticAttachmentImportPaths,
  deletableFiles: RA<PartialUploadableFileSpec>
): Promise<RA<PartialUploadableFileSpec>> {
  const baseTable = staticAttachmentImportPaths[staticKey].baseTable;
  const path = `${AttachmentMapping[baseTable].relationship}.${staticAttachmentImportPaths[staticKey].restPath}`;
  const relatedAttachments = filterArray(
    deletableFiles.map((deletable) =>
      deletable.status?.type === 'matched' ? deletable.attachmentId : undefined
    )
  );
  const reconstructingQueryResource = generateInQueryResource(
    baseTable,
    path,
    relatedAttachments,
    'Batch Attachment Upload'
  );
  const queryResults = await validationPromiseGenerator(
    reconstructingQueryResource
  );
  return deletableFiles.map((deletable) => {
    if (deletable.status?.type !== 'matched') return deletable;
    const matchedId = deletable.status.id;
    const foundInQueryResult = queryResults.find(
      ({ targetId, restResult: [attachmentId] }) =>
        targetId === matchedId && attachmentId === deletable.attachmentId
    );
    return {
      ...deletable,
      attachmentId: foundInQueryResult?.restResult[0] as number,
      status:
        foundInQueryResult === undefined
          ? ({ type: 'success', successType: 'uploaded' } as const)
          : deletable.status.type === 'matched'
          ? ({
              type: 'cancelled',
              reason: 'rollbackInterruption',
            } as const)
          : deletable.status,
    };
  });
}
export async function reconstructUploadingAttachmentSpec(
  staticKey: keyof typeof staticAttachmentImportPaths,
  uploadableFiles: RA<PartialUploadableFileSpec>
): Promise<RA<PartialUploadableFileSpec>> {
  const baseTable = staticAttachmentImportPaths[staticKey].baseTable;
  const pathToAttachmentLocation = `${AttachmentMapping[baseTable].relationship}.attachment.attachmentLocation`;
  const filteredAttachmentLocations = filterArray(
    uploadableFiles.map((uploadable) =>
      uploadable.status?.type === 'matched'
        ? uploadable.uploadTokenSpec?.attachmentlocation
        : undefined
    )
  );
  const reconstructingQueryResource = generateInQueryResource(
    baseTable,
    pathToAttachmentLocation,
    filteredAttachmentLocations,
    'Batch Attachment Upload',
    [
      {
        path: `${AttachmentMapping[baseTable].relationship}.${staticAttachmentImportPaths[staticKey].restPath}`,
        isDisplay: true,
        id: queryFieldFilters.any.id,
      },
    ]
  );
  const queryResults = await validationPromiseGenerator(
    reconstructingQueryResource
  );
  return uploadableFiles.map((uploadable) => {
    if (uploadable.status?.type !== 'matched') return uploadable;
    const matchedId = uploadable.status.id;
    const foundInQueryResult = queryResults.find(
      ({ targetId, restResult: [_, attachmentLocation] }) =>
        typeof attachmentLocation === 'string' &&
        targetId === matchedId &&
        attachmentLocation.toString() ===
          uploadable.uploadTokenSpec!.attachmentlocation
    );

    return {
      ...uploadable,
      attachmentId: foundInQueryResult?.restResult[0] as number,
      status:
        typeof foundInQueryResult === 'object'
          ? ({ type: 'success', successType: 'uploaded' } as const)
          : uploadable.status.type === 'matched'
          ? ({
              type: 'cancelled',
              reason: 'uploadInterruption',
            } as const)
          : uploadable.status,
    };
  });
}

export const keyLocalizationMapAttachment = {
  incorrectFormatter: attachmentsText.incorrectFormatter(),
  noFile: attachmentsText.noFile(),
  uploaded: commonText.uploaded(),
  deleted: attachmentsText.deleted(),
  alreadyUploaded: attachmentsText.alreadyUploaded(),
  alreadyDeleted: attachmentsText.alreadyDeleted(),
  skipped: attachmentsText.skipped(),
  cancelled: attachmentsText.cancelled(),
  matchError: attachmentsText.matchError(),
  noAttachments: attachmentsText.noAttachments(),
  uploadInterruption: attachmentsText.frontEndInterruption({
    action: wbText.upload(),
  }),
  rollbackInterruption: attachmentsText.frontEndInterruption({
    action: wbText.rollback(),
  }),
  errorReadingFile: attachmentsText.errorReadingFile(),
  attachmentServerUnavailable: attachmentsText.attachmentServerUnavailable(),
  saveConflict: formsText.saveConflict(),
  unhandledFatalResourceError: attachmentsText.unhandledFatalResourceError(),
  nothingFound: formsText.nothingFound(),
  noMatch: attachmentsText.noMatch(),
  multipleMatches: attachmentsText.multipleMatches(),
  correctlyFormatted: attachmentsText.correctlyFormatted(),
  userStopped: attachmentsText.stoppedByUser(),
} as const;

export function resolveAttachmentStatus(
  attachmentStatus: AttachmentStatus
): string {
  if ('reason' in attachmentStatus) {
    const reason = keyLocalizationMapAttachment[attachmentStatus.reason];
    return commonText.colonLine({
      label: keyLocalizationMapAttachment[attachmentStatus.type],
      value: reason,
    });
  }
  return attachmentStatus.type === 'success'
    ? keyLocalizationMapAttachment[attachmentStatus.successType]
    : '';
}
