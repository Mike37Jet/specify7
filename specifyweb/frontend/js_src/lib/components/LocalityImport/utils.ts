import type { LocalizedString } from 'typesafe-i18n';

import { commonText } from '../../localization/common';
import { localityText } from '../../localization/locality';
import { f } from '../../utils/functools';
import type { IR, RA, RR } from '../../utils/types';
import { tables } from '../DataModel/tables';
import type { Tables } from '../DataModel/types';
import { resolveBackendParsingMessage } from '../WorkBench/resultsParser';
import type { LocalityImportHeader, LocalityImportTaskStatus } from './types';

export const localityImportAcceptedLocalityFields: RA<
  Lowercase<keyof Tables['Locality']['fields']>
> = ['guid', 'datum', 'latitude1', 'longitude1'];

export const localityImportAcceptedHeaders = f.store(
  () =>
    new Set([
      ...localityImportAcceptedLocalityFields,
      ...tables.GeoCoordDetail.literalFields
        .map(({ name }) => name.toLowerCase())
        .filter((header) => header !== 'locality'),
    ])
);

export const localityImportRequiredHeaders = new Set<LocalityImportHeader>([
  'guid',
]);

export const localityImportStatusLocalization: RR<
  LocalityImportTaskStatus,
  LocalizedString
> = {
  PARSING: localityText.localityImportParsing(),
  PENDING: localityText.localityImportStarting(),
  PROGRESS: localityText.localityImportProgressing(),
  FAILED: localityText.localityImportFailed(),
  ABORTED: localityText.localityImportCancelled(),
  SUCCEEDED: localityText.localityImportSucceeded(),
};

export function resolveImportLocalityErrorMessage(
  key: string,
  payload: IR<unknown>
): LocalizedString {
  const baseParseResults = resolveBackendParsingMessage(key, payload);

  if (baseParseResults !== undefined) {
    return baseParseResults;
  } else if (key === 'guidHeaderNotProvided') {
    return localityText.guidHeaderNotProvided();
  } else if (key === 'noLocalityMatchingGuid') {
    return localityText.noLocalityMatchingGuid({
      guid: payload.guid as string,
    });
  } else if (key === 'multipleLocalitiesWithGuid') {
    return localityText.multipleLocalitiesWithGuid({
      guid: payload.guid as string,
      localityIds: (payload.localityIds as RA<number>).join(', '),
    });
  } else {
    return commonText.colonLine({
      label: key,
      value:
        Object.keys(payload).length === 0 ? '' : `${JSON.stringify(payload)}`,
    });
  }
}
