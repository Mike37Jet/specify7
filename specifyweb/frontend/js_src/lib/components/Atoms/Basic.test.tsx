import { render } from '@testing-library/react';
import React from 'react';

import { commonText } from '../../localization/common';
import { mount } from '../../tests/utils';
import { ErrorMessage, Input } from './Basic';

test('ErrorMessage renders without errors', () => {
  const { asFragment } = render(
    <ErrorMessage>{commonText('title')}</ErrorMessage>,
    {}
  );

  expect(asFragment()).toMatchSnapshot();
});

test('Input.Text emits onValueChange', async () => {
  const handleValueChange = jest.fn();
  const { getByRole, user } = mount(
    <Input.Text name="test-input" value="" onValueChange={handleValueChange} />,
    {}
  );

  const input = getByRole('textbox');
  await user.type(input, 'abc');
  expect(handleValueChange).toHaveBeenCalledTimes(3);
  expect(handleValueChange).toHaveBeenNthCalledWith(1, 'a');
  expect(handleValueChange).toHaveBeenNthCalledWith(2, 'b');
  expect(handleValueChange).toHaveBeenNthCalledWith(3, 'c');
});
