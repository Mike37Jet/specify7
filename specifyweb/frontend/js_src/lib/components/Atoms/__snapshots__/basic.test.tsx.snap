// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage renders without errors 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-2 p-2 text-white bg-red-500 rounded"
    role="alert"
  >
    Title
  </div>
</DocumentFragment>
`;